ui/notifications.ts

```diff
@@ -0,0 +1,180 @@
+  ${notifications.length > 0 ? notifications.reverse().map(value => singleNotification(value)).join('') : empty}
+</div>`;
+
+export class NotificationManager {
+
+    private static NOTIFICATION_KEY = 'service_notifications';
+
+    private notifications: Notification[] = [];
+
+    getNotifications(): Notification[] {
+        return this.notifications;
+    }
+
+    dismissNotification(id: number) {
+        const index = this.notifications.findIndex(v => v.id === id);
+        if (index > -1) {
+            this.notifications.splice(index, 1);
+        }
+        this.toCache(this.notifications);
+        console.log(this.notifications);
+    }
+
+    mergeToCache(notifications: Notification[]) {
+        const data = this.fromCache().concat(notifications);
+        this.toCache(data);
+        localStorage.setItem(NotificationManager.NOTIFICATION_KEY, JSON.stringify(data));
+        this.notifications = data;
+    }
+
+    clearAll() {
+        this.notifications = [];
+        this.toCache(this.notifications);
+    }
+
+    getNextId(): number {
+        const ids = this.notifications.map(value => value.id)
+            .sort((a, b) => b - a);
+
+        let nextId: number;
+        if (ids.length > 0) {
+            nextId = ids[0] + 1;
+        } else {
+            nextId = 1;
+        }
+        return nextId;
+    }
+
+    private toCache(notifications: Notification[]) {
+        localStorage.setItem(NotificationManager.NOTIFICATION_KEY, JSON.stringify(notifications));
+    }
+
+    private fromCache(): Notification[] {
+        const stored = localStorage.getItem(NotificationManager.NOTIFICATION_KEY);
+        let cached: Notification[] = [];
+        if (stored) {
+            cached = JSON.parse(stored);
+        }
+        return cached;
+    }
+}
+
+export const notificationManager = new NotificationManager();
+
+function updateNotificationMessages(notifications: Notification[], elements: JQuery<HTMLElement>) {
+    for (const notification of notifications) {
+        const element = elements.filter(`#notification-${notification.id}`);
+        if (element.html()) {
+            continue;
+        }
+        $('.notification-area').prepend(singleNotification(notification));
+    }
+
+}
+export function updateNotifications() {
+    const notifications = notificationManager.getNotifications();
+    const badge = $('#notification-badge');
+    const messages = $('#notification-messages');
+
+    badge.html(notificationBadge(notifications));
+
+    const elements = $('.notification');
+    if (elements.length === 0) {
+        messages.html(notificationMessages(notifications));
+    } else {
+        updateNotificationMessages(notifications, elements);
+    }
+
+}
+export function setupNotificationHandlers() {
+    const $notification = $('.notification');
+
+    const $clear = $('#notifications-clear-all');
+    $notification.off('click');
+
+    $clear.off('click');
+
+    $notification.on('click', function (event: JQuery.Event) {
```

Comment:
![2019-05-20-172722_339x426_scrot](https://user-images.githubusercontent.com/1658405/58033738-51ecf180-7b25-11e9-9e93-d614c11c8f85.png)

So I have some comments here.

1. I see that the notification is being dismissed when it's clicked. I would rather there is a specific button. Like an [X] on the top right of each notification that dismisses it. Clicking anywhere else should not dismiss it. Perhaps there should also be a tooltip on top of that [X].
2. The garbage bin button on the top right is a bit too big. Should be a tiny tiny bit smaller. And most importantly it should:
    - have a tooltip that appears when hovering over it that will say that clicking will dismiss all open notification
    - have a confirmation menu. Clicking the button should not immediately remove all notifications but should ask you if you really want all active notifications to be removed.

Reply:
Shouldn't be too much effort to update it. Thanks.

Reply:
Okay this is done. Tested it. Thanks.